ports/espressif/common-hal/espulp/ULP.c

```diff
@@ -24,49 +24,104 @@
 #elif defined(CONFIG_IDF_TARGET_ESP32S3)
 #include "esp32s3/ulp.h"
 #include "esp32s3/ulp_riscv.h"
+#define ULP_COPROC_RESERVE_MEM (CONFIG_ESP32S3_ULP_COPROC_RESERVE_MEM)
 #endif
 
 // To-do idf v5.0: remove following include
 #include "soc/rtc_cntl_reg.h"
 
-void common_hal_coproc_run(coproc_coproc_obj_t *self) {
+STATIC bool ulp_used = false;
+STATIC uint32_t pins_used = 0;
+
+void espulp_reset(void) {
+    // NOTE: This *doesn't* disable the ULP. It'll keep running even when CircuitPython isn't.
+    ulp_used = false;
+}
+
+void common_hal_espulp_ulp_run(espulp_ulp_obj_t *self, uint32_t *program, size_t length, uint32_t pin_mask) {
+    if (length > ULP_COPROC_RESERVE_MEM) {
+        mp_raise_ValueError(translate("Program too long"));
+    }
     if (GET_PERI_REG_MASK(RTC_CNTL_ULP_CP_TIMER_REG, RTC_CNTL_ULP_CP_SLP_TIMER_EN)) {
         mp_raise_RuntimeError(translate("Already running"));
     }
 
-    ulp_riscv_load_binary(self->buf, self->buf_len);
-    m_free(self->buf);
-    self->buf = (uint8_t *)RTC_SLOW_MEM;
+    if (pin_mask >= (1 << 22)) {
+        mp_raise_ValueError(translate("Pins 21+ not supported from ULP"));
+    }
+
+    for (uint8_t i = 0; i < 32; i++) {
+        if ((pin_mask & (1 << i)) != 0 && !pin_number_is_free(i)) {
+            mp_raise_ValueError_varg(translate("%q in use"), MP_QSTR_Pin);
+        }
+    }
+
+    for (uint8_t i = 0; i < 32; i++) {
+        if ((pin_mask & (1 << i)) != 0) {
+            claim_pin_number(i);
+            never_reset_pin_number(i);
+        }
+    }
+    pins_used = pin_mask;
+
 
+    ulp_riscv_load_binary((const uint8_t *)program, length);
+    ulp_set_wakeup_period(0, 20000);
     ulp_riscv_run();
 }
 
-void common_hal_coproc_halt(coproc_coproc_obj_t *self) {
-    self->buf = (uint8_t *)m_malloc(self->buf_len, false);
-    memcpy(self->buf, (uint8_t *)RTC_SLOW_MEM, self->buf_len);
-
+void common_hal_espulp_ulp_halt(espulp_ulp_obj_t *self) {
     // To-do idf v5.0: use following functions
     // ulp_riscv_timer_stop();
     // ulp_riscv_halt();
 
-    // stop the ulp timer
+    // stop the ulp timer so that is doesn't restart the cpu
     CLEAR_PERI_REG_MASK(RTC_CNTL_ULP_CP_TIMER_REG, RTC_CNTL_ULP_CP_SLP_TIMER_EN);
+
     // suspends the ulp operation
     SET_PERI_REG_MASK(RTC_CNTL_COCPU_CTRL_REG, RTC_CNTL_COCPU_DONE);
+
     // resets the processor
     SET_PERI_REG_MASK(RTC_CNTL_COCPU_CTRL_REG, RTC_CNTL_COCPU_SHUT_RESET_EN);
+
+    // Release pins we were using.
+    for (uint8_t i = 0; i < 32; i++) {
+        if ((pins_used & (1 << i)) != 0) {
+            reset_pin_number(i);
+        }
+    }
 }
 
-mp_obj_t common_hal_coproc_memory(coproc_coproc_obj_t *self) {
-    return (self->coproc_memory) ? MP_OBJ_FROM_PTR(self->coproc_memory) : mp_const_none;
+void common_hal_espulp_ulp_construct(espulp_ulp_obj_t *self) {
+    if (ulp_used) {
+        mp_raise_ValueError_varg(translate("%q in use"), MP_QSTR_ULP);
+    }
+    self->inited = true;
```

Comment:
Might want to infer the ULP state from registers.

Reply:
This was deliberate because the ULP might already be running when the object is created.